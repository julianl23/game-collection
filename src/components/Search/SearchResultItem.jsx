import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { theme } from "styled-tools";
import ReactCSSTransitionGroup from "react-addons-css-transition-group";

import Button from "../Button";
import AddItemDetailView from "./AddItemDetailView";
import { PlusSquare } from "../Icons";
import Snackbar from "../Snackbar";

const Item = styled.li`
  margin-bottom: 48px;
`;

const AddButton = styled(Button)`
  display: flex;
  justify-content: center;
  align-items: center;

  svg {
    margin-right: 5px;
  }
`;

const SearchResultInfo = styled.div`
  display: flex;
  margin-bottom: 20px;
`;

const ContentSection = styled.div`
  margin-left: 13px;
`;

const Companies = styled.p`
  margin: 0;
`;

const Title = styled.span`
  font-size: 16px;
  font-weight: 500;
  line-height: 19px;
  color: ${theme("deepPurple")};
  margin-bottom: 7px;
  display: inline-block;
`;

const Platforms = styled.p`
  margin: 20px 0 0;
`;

class SearchResultItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      expanded: false,
      itemAdded: false,
    };
  }

  renderCompanyList = (companies, prefix) => {
    const companyNames = companies.map(company => company.name);
    let companyNameString;

    if (!companies.length) {
      return null;
    }

    if (companies.length > 1) {
      const otherCount = companies.length - 2;
      const othersString = `${otherCount} ${otherCount > 1 && "s"}`;
      companyNameString = `${companyNames[0]}, ${
        companyNames[1]
      }, and ${othersString}`;
    } else {
      companyNameString = companyNames.join(", ");
    }

    return (
      <React.Fragment>
        {prefix} {companyNameString}
      </React.Fragment>
    );
  };

  renderPlatformList = platforms => {
    const platformNames = platforms.map(platform => platform.name);
    let platformsString;

    if (!platforms.length) {
      return null;
    }

    if (platforms.length > 2) {
      const otherCount = platforms.length - 2;
      const othersString = `${otherCount} other${otherCount > 1 ? "s" : ""}`;
      platformsString = `${platformNames[0]}, ${
        platformNames[1]
      }, and ${othersString}`;
    } else {
      platformsString = platformNames.join(", ");
    }

    return <Platforms>Platforms: {platformsString}</Platforms>;
  };

  handleToggleAddView = () => {
    const { expanded } = this.state;
    this.setState({
      expanded: !expanded,
    });
  };

  handleItemAdded = () => {
    this.setState(
      {
        itemAdded: true,
      },
      () => {
        setTimeout(() => {
          this.setState({
            itemAdded: false,
          });
        }, 3000);
      }
    );

    this.handleToggleAddView();
  };

  render() {
    const { game } = this.props;
    const { expanded, itemAdded } = this.state;

    const { cover, title, developer, publisher, platforms } = game;

    return (
      <Item>
        <SearchResultInfo>
          <img src={cover.url} alt={`Cover of ${title}`} />
          <ContentSection>
            <Title>{title}</Title>
            <Companies>
              {this.renderCompanyList(developer, "Developed by")}
              <br />
              {this.renderCompanyList(publisher, "Published by")}
            </Companies>

            {this.renderPlatformList(platforms)}
          </ContentSection>
        </SearchResultInfo>
        <ReactCSSTransitionGroup
          transitionName="snackbar-fade"
          transitionAppear
          transitionAppearTimeout={250}
          transitionEnterTimeout={250}
          transitionLeave
          transitionLeaveTimeout={500}
        >
          {itemAdded && (
            <Snackbar key={0}>
              <em>{title}</em> has been added to your collection.
            </Snackbar>
          )}
        </ReactCSSTransitionGroup>
        {expanded && (
          <AddItemDetailView
            game={game}
            handleItemAdded={this.handleItemAdded}
            handleToggleAddView={this.handleToggleAddView}
          />
        )}
        {!expanded && (
          <AddButton block onClick={this.handleToggleAddView}>
            <PlusSquare />Add to my collection
          </AddButton>
        )}
      </Item>
    );
  }
}

SearchResultItem.propTypes = {
  game: PropTypes.object.isRequired,
};

export default SearchResultItem;
