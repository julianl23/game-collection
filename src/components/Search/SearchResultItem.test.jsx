import React from "react";
import TestRenderer from "react-test-renderer";

import SearchResultItem from "./SearchResultItem";

const mockGame = {
  _id: "123456",
  platforms: [
    {
      _id: "123456",
      name: "Sega GameCast",
    },
  ],
  cover: {
    url: "http://image.com/image.jpg",
  },
  title: "Sonic Rides Again",
  developer: [{ name: "Sonic Team" }],
  publisher: [{ name: "Sega" }],
};

describe("SearchResultItem", () => {
  it("renders correctly", () => {
    const instance = TestRenderer.create(<SearchResultItem game={mockGame} />);
    expect(instance.toJSON()).toMatchSnapshot();
  });
});
