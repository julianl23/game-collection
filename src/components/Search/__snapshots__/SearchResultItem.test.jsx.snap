// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultItem renders correctly 1`] = `
.c7 {
  height: 36px;
  background: #322F6F;
  border-radius: 3px;
  font-size: 14px;
  font-weight: 300;
  line-height: 16px;
  color: #fff;
  cursor: pointer;
  display: block;
  width: 100%;
}

.c0 {
  margin-bottom: 48px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 svg {
  margin-right: 5px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 20px;
}

.c2 {
  margin-left: 13px;
}

.c4 {
  margin: 0;
}

.c3 {
  font-size: 16px;
  font-weight: 500;
  line-height: 19px;
  margin-bottom: 7px;
  display: inline-block;
}

.c5 {
  margin: 20px 0 0;
}

<li
  className="c0"
>
  <div
    className="c1"
  >
    <img
      alt="Cover of Sonic Rides Again"
      src="http://image.com/image.jpg"
    />
    <div
      className="c2"
    >
      <span
        className="c3"
      >
        Sonic Rides Again
      </span>
      <p
        className="c4"
      >
        Developed by
         
        Sonic Team
        <br />
        Published by
         
        Sega
      </p>
      <p
        className="c5"
      >
        Platforms: 
        Sega GameCast
      </p>
    </div>
  </div>
  <span />
  <button
    className="c6 c7"
    onClick={[Function]}
    size="normal"
    type="button"
  >
    <svg
      fill="#fff"
      height={22}
      viewBox="0 0 448 512"
      width={20}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M352 240v32c0 6.6-5.4 12-12 12h-88v88c0 6.6-5.4 12-12 12h-32c-6.6 0-12-5.4-12-12v-88h-88c-6.6 0-12-5.4-12-12v-32c0-6.6 5.4-12 12-12h88v-88c0-6.6 5.4-12 12-12h32c6.6 0 12 5.4 12 12v88h88c6.6 0 12 5.4 12 12zm96-160v352c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V80c0-26.5 21.5-48 48-48h352c26.5 0 48 21.5 48 48zm-48 346V86c0-3.3-2.7-6-6-6H54c-3.3 0-6 2.7-6 6v340c0 3.3 2.7 6 6 6h340c3.3 0 6-2.7 6-6z"
      />
    </svg>
    Add to my collection
  </button>
</li>
`;
